### PR TITLE
Extend modelbuilders test set

### PR DESCRIPTION
**Pull request by avolkov-intel**
## Description

Since we plan an update to internal modelbuilders layout by introducing hybrid layout (https://github.com/uxlfoundation/oneDAL/pull/3455) we need to extend test set.
In the new layouts we would create full binary tree layout for levels <= 10 but after that we will store nodes in dense format. To check, that this implementation doesn't have any bugs we need to add tests where deep trees are used.

This test introduces new tests where trees with depth >=12 or >=15 created.

---

<details>
<summary>Checklist:</summary>

**Completeness and readability**

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with updates and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

</details>

### CODE REVIEW

**Comment by david-cortes-intel** [comment]
Wouldn't it make more sense to have tests like this on the oneDAL repository?

File: tests/test_model_builders.py [hunk 1]
```diff
@@ -106,81 +107,174 @@ def force_shap_predict(model, X):
         return model._predict_classification(X, fptype, "computeClassLabels", True, False)
 
 
+# ============================================================================
+# Tree depth utility functions for deep trees testing
+# ============================================================================
+def compute_xgb_tree_depth(booster: xgb.Booster) -> int:
```
Comment (david-cortes-intel) [inline tests/test_model_builders.py:113]: Since the idea is to test some internal detail of the implementation that doesn't anyhow depend on the input converter, how about checking only a single GBT library? Perhaps a random forest from sklearn converted with treelite could be used.

Comment (avolkov-intel) [reply tests/test_model_builders.py:113]: We actually have some existing tests related to xgboost and catboost including SHAP tests, it would require much more effort to create tests using RandomForest and TreeLite compared to updating existing tests

Comment (david-cortes-intel) [reply tests/test_model_builders.py:113]: Then how about leaving serialized model objects instead of re-creating them every time? These tests are already having issues in CIs due to taking too long to execute, and some are currently skipped due to timings.

**Comment by ethanglaser** [comment]
Please address linter fail

File: tests/test_model_builders.py (hunk 1 above)
Comment (avolkov-intel) [reply tests/test_model_builders.py:113]: I think this would make the testing more complicated, not sure if we use serialized models in other tests. Although, I agree that modelbuilders tests take quite a lot of time, the relative slowdown after introducing tests in this PR is small. That's why let's create a follow up for speeding up this set of tests

Comment (david-cortes-intel) [reply tests/test_model_builders.py:113]: From what I can see in the CI jobs here compared to those of recent commits, it adds roughly 1 additional minute in the CI servers.
